multicluster/controllers/multicluster/label_identity_export_controller.go

```diff
@@ -0,0 +1,296 @@
+	for _, label := range deletedLabels {
+		existingLabelImport := &mcsv1alpha1.LabelIdentityImport{}
+		resNamespaced := types.NamespacedName{
+			Name:      hashLabelIdentity(label),
+			Namespace: resExport.Namespace,
+		}
+		err := r.Client.Get(ctx, resNamespaced, existingLabelImport)
+		if err != nil && !apierrors.IsNotFound(err) {
+			klog.ErrorS(err, "failed to get LabelIdentityImport for stale label", "label", label)
+			reconcileFailedLabels.Insert(label)
+			continue
+		}
+		if err := r.Client.Delete(ctx, existingLabelImport, &client.DeleteOptions{}); err != nil {
+			klog.ErrorS(err, "failed to delete LabelIdentityImport for stale label", "label", label)
+			reconcileFailedLabels.Insert(label)
+			continue
+		}
+		// Delete caches for the label and release the ID assigned for the label
+		id, _ := r.labelsToID[label]
+		r.allocator.release(id)
+		delete(r.labelsToID, label)
+		delete(r.labelsToClusters, label)
+		if clusterLabels, ok := r.clusterToLabels[clusterID]; ok {
+			clusterLabels.Delete(label)
+		}
+	}
+}
+
+// handleLabelIdentitiesDelete creates LabelIdentityImports of label identities that are added
+// in the ClusterSet. Note that the ID of a label identity is only allocated and stored if the
+// creation of its LabelIdentityImport succeeded.
+func (r *LabelIdentityExportReconciler) handleLabelIdentitiesAdd(ctx context.Context,
+	addedLabels []string, reconcileFailedLabels sets.String, clusterID string, resExport *mcsv1alpha1.ResourceExport) {
+	for _, label := range addedLabels {
+		id, err := r.allocator.allocate()
+		if err != nil {
+			klog.ErrorS(err, "failed to allocate ID for new label", "label", label)
+			reconcileFailedLabels.Insert(label)
+			continue
+		}
+		labelIdentityImport := getLabelIdentityImport(label, resExport.Namespace, id)
+		if err := r.Client.Create(ctx, labelIdentityImport, &client.CreateOptions{}); err != nil {
+			// TODO: check apierrors.IsAlreadyExists(err) to see if it is a hash collision?
+			klog.ErrorS(err, "failed to create LabelIdentityImport for new label", "label", label)
+			r.allocator.release(id)
+			reconcileFailedLabels.Insert(label)
+			continue
+		}
+		r.labelsToID[label] = id
+		if clusters, ok := r.labelsToClusters[label]; ok {
+			clusters.Insert(clusterID)
+		} else {
+			r.labelsToClusters[label] = sets.NewString(clusterID)
+		}
+		if labels, ok := r.clusterToLabels[clusterID]; ok {
+			labels.Insert(label)
+		} else {
+			r.clusterToLabels[clusterID] = sets.NewString(label)
+		}
+	}
+}
+
+func getLabelIdentityImport(label, ns string, id uint32) *mcsv1alpha1.LabelIdentityImport {
+	return &mcsv1alpha1.LabelIdentityImport{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      hashLabelIdentity(label),
+			Namespace: ns,
+		},
+		Spec: mcsv1alpha1.LabelIdentityImportSpec{
+			Label: label,
+			ID:    id,
+		},
+	}
+}
+
+func (r *LabelIdentityExportReconciler) deleteResourceExport(resExport *mcsv1alpha1.ResourceExport) (ctrl.Result, error) {
```

Comment:
This is confused, you name it `deleteResourceExport` but call `update`? and this method is not used, can it be removed?

Reply:
Because originally I was using finalizers and the ResourceExport will be deleted once the finalizer is removed. But this is not the case anymore. Removed the code